dlt/destinations/impl/filesystem/filesystem.py

```diff
@@ -455,3 +461,97 @@ def get_stored_schema(self) -> Optional[StorageSchemaInfo]:
+                        schema = schema.set(i, adjusted_field)
+                        break  # if type matches type check, do not do other type checks
+            return schema
+
+        def ensure_delta_compatible_arrow_table(table: pa.table) -> pa.Table:
+            """Returns Arrow table compatible with Delta table format.
+
+            Casts table schema to replace data types not supported by Delta.
+            """
+            ARROW_TO_DELTA_COMPATIBLE_ARROW_TYPE_MAP = {
+                # maps type check function to type factory function
+                pa.types.is_null: pa.string(),
+                pa.types.is_time: pa.string(),
+                pa.types.is_decimal256: (
+                    pa.string()
+                ),  # pyarrow does not allow downcasting to decimal128
+            }
+            adjusted_schema = adjust_arrow_schema(
+                table.schema, ARROW_TO_DELTA_COMPATIBLE_ARROW_TYPE_MAP
+            )
+            return table.cast(adjusted_schema)
+
+        def get_delta_write_mode(write_disposition: TWriteDisposition) -> str:
+            """Translates dlt write disposition to Delta write mode."""
+            if write_disposition in ("append", "merge"):  # `merge` disposition resolves to `append`
+                return "append"
+            elif write_disposition == "replace":
+                return "overwrite"
+            else:
+                raise ValueError(
+                    "`write_disposition` must be `append`, `replace`, or `merge`,"
+                    f" but `{write_disposition}` was provided."
+                )
+
+        # throws warning for `s3` protocol: https://github.com/delta-io/delta-rs/issues/2460
+        # TODO: upgrade `deltalake` lib after https://github.com/delta-io/delta-rs/pull/2500
+        # is released
+        write_deltalake(  # type: ignore[call-overload]
+            table_or_uri=path,
+            data=ensure_delta_compatible_arrow_table(table),
+            mode=get_delta_write_mode(write_disposition),
+            schema_mode="merge",  # enable schema evolution (adding new columns)
+            storage_options=self._deltalake_storage_options,
+            engine="rust",  # `merge` schema mode requires `rust` engine
+        )
+
+    @property
+    def _deltalake_storage_options(self) -> Dict[str, str]:
+        """Returns dict that can be passed as `storage_options` in `deltalake` library."""
+        # TODO: implement method properly and remove hard-coded values.
+        # `delta-rs` does not currently (version 0.17.4) support `gdrive` protocol
+        storage_options = None
+        if self.config.protocol == "s3":
+            from dlt.common.configuration.specs import AwsCredentials
+
+            assert isinstance(self.config.credentials, AwsCredentials)
+            storage_options = self.config.credentials.to_session_credentials()
+            storage_options["AWS_REGION"] = self.config.credentials.region_name
```

Comment:
The `deltalake` library requires that `AWS_REGION` is provided. We need to add it to `DLT_SECRETS_TOML` under `[destination.filesystem.credentials]` to make `s3` tests pass on CI.